src/compositor/shells/xdg_v6.rs

```diff
@@ -13,7 +15,33 @@ impl XdgV6 {
     }
 }
 
-impl XdgV6ShellHandler for XdgV6 {}
+impl XdgV6ShellHandler for XdgV6 {
+    fn move_request(&mut self,
+                    compositor: CompositorHandle,
+                    surface: SurfaceHandle,
+                    shell_surface: XdgV6ShellSurfaceHandle,
+                    event: &MoveEvent) {
+        with_handles!([(compositor: {compositor})] => {
+            let server: &mut Server = compositor.data.downcast_mut().unwrap();
+            let ref mut seat = server.seat;
+            let ref mut cursor = server.cursor;
+
+            seat.focused = seat.focused.take().and_then(|view| {
```

Comment:
You could do this better like this:

```rust
            if let Some(ref mut view) = seat.focused {
                let shell: Shell = shell_surface.into();
                let action = &mut seat.action;
                if view.shell == shell {
                    with_handles!([(cursor: {cursor})] => {
                        let (lx, ly) = cursor.coords();
                        let Origin { x: shell_x, y: shell_y } = view.origin;
                        let (view_sx, view_sy) = (lx - shell_x as f64, ly - shell_y as f64);
                        let start = Origin::new(view_sx as _, view_sy as _);
                        *action = Some(Action::Moving { start: start });
                    }).unwrap();
                }
            }
```

Note that we get around the closure borrowing the entire struct by taking a mutable reference to the field we want to mutate.

Reply:
Yeah this will be needed in multiple places as well, so I'll probably make it a server method.